crates/ibc/src/core/ics24_host/identifier/validate.rs

```diff
@@ -74,7 +61,6 @@ pub fn validate_prefix_length(
     if max_id_length < 22 {
```

Comment:
This is a valid error. This is equivalent to the following error,
```rs
Error::InvalidLength {
  id: format!("{}-{}", prefix, u64::MAX),
  min: min_id_length,
  max: max_id_length,
}
```

Reply:
Maybe we need an `InvalidPrefixLength` error variant to distinguish properly.

Reply:
> There already is check for `format!("{}-{}", prefix, u64::MAX)` being too long.

Not anymore. We use `validate_identifier_length(prefix, min_id_length.saturating_sub(21), max_id_length.saturating_sub(21))` directly.

`max_id_length < 22` makes sure two things. (1) the prefix is not empty (2) `max_id_length - 21` doesn't underflow.

Reply:
> > There already is check for `format!("{}-{}", prefix, u64::MAX)` being too long.
>
> Not anymore. We use `validate_identifier_length(prefix, min_id_length.saturating_sub(21), max_id_length.saturating_sub(21))` directly.

Right, that’s what I meant.  If max_id_length < 22, that check will fail.

> `max_id_length < 22` makes sure two things. (1) the prefix is not empty

There’s explicit check for that at the beginning of the function:

    if prefix.is_empty() {
        return Err(Error::InvalidPrefix {
            prefix: prefix.into(),
        });
    }

> (2) `max_id_length - 21` (used as `max_id_length.checked_sub(21)`) doesn't underflow.

        max_id_length.saturating_sub(21),

handles underflow just fine.

The entire validate_identifier_length could be this:

```rust
pub fn validate_prefix_length(
    prefix: &str,
    min_id_length: u64,
    max_id_length: u64,
) -> Result<(), Error> {
    // Prefix must be at least `min_id_length - 2` characters long since the
    // shortest identifier we can construct is `{prefix}-0` which extends prefix
    // by two characters.
    let min = min_id_length.saturating_sub(2).max(1);
    // Prefix must be at most `max_id_length - 21` characters long since the
    // longest identifier we can construct is `{prefix}-{u64::MAX}` which
    // extends prefix by 21 characters.
    let max = max_id_length.saturating_sub(21);
    if (min..=max).contains(&(prefix.len() as u64)) {
        Ok(())
    } else {
        Err(Error::InvalidLength { /* ... */ })
    }
}
```

* If `max_id_length < min_id_length` than `max(max_id_length - 21, 0)
  < max(min_id_length - 2, 1)` and the `min..=max` range will be empty.
* If prefix is empty, than `prefix.len() < max(min_id_length - 2, 1)`.
* If `max_id_length < 22` than `max_id_length - 22 == 0` and since
  prefix cannot be empty (per previous point) `max_id_length - 22 <
  prefix.len()`.

Reply:
Here’s a draft PR with that change: https://github.com/cosmos/ibc-rs/pull/960

Reply:
This looks excellent :ok_hand:

You optimized one of my old commits https://github.com/cosmos/ibc-rs/commit/69301f8f8c30fd3d26cfb2ce1b042ce30a0cc652 that I reverted.